### ISSUE TITLE
Real-time kernel stack usage monitoring

### ISSUE THREAD

**Issue by milabs** [milabs/kmod_hooking_sct#1]
Let's do the real-time kernel stack usage monitoring tool (JS + kernel-side module) like that:

http://techblog.netflix.com/2014/11/nodejs-in-flames.html
http://www.brendangregg.com/blog/2014-09-17/node-flame-graphs-on-linux.html

@alexandernst

**Comment by alexandernst** [comment]
@milabs Really interesting indeed. It could be used to improve the kernel itself. Also, nice project you started! It reminds me a lot of Monks :smile: (I really should fix https://github.com/alexandernst/monks/issues/16 and make a release!)

I can do the node/html part that would receive/render the information that the kernel module can obtain.
I have been willing to play with some plot libs, and this sounds just like the right project for it :)

**Comment by milabs** [comment]
@alexandernst Let's fork the repo and trying to get it work? I'll take a kernel part and you'll take the plotting/rendering one? It would be nice to see how our systems kernel stack's lives in real-time :)

**Comment by alexandernst** [comment]
@milabs Yup! Sounds great! How do you plan sending the data from the kernel to userland? And more specifically, to NodeJS?

**Comment by milabs** [comment]
@alexandernst Don't know how to do it right. Thinking about ring buffer of events which can be mapped on a process or may be in-kernel table for each of the syscall with information about min/max stack usage values and some others being landed to the user by request.

**Comment by alexandernst** [comment]
@milabs Remember how I did it in Monks? I used netlink to sent a char containing a custom serialized data (in the form of (num_chunks)(size)(chunk)(size)(chunk).... ): https://github.com/alexandernst/monks/blob/master/monks_kmodule/netlink.c#L84

and then I used https://github.com/alexandernst/monks/blob/master/monks/netlink.c#L31 to read to userland.
Maybe we could do the same for this?

BTW, sorry for taking so long to reply. There is a ton of work to do at work lately :(

**Comment by milabs** [comment]
@alexandernst Don't you think that it was slow and a little bit inefficient? :)

**Comment by alexandernst** [comment]
@milabs It probably is, but I couldn't find any other (easy) method :P

**Comment by milabs** [comment]
@alexandernst check the https://github.com/milabs/stamina repo
